File: tools/java_api/src/main/java/com/kuzudb/Connection.java
package com.kuzudb;

import java.util.Map;

/**
 * Connection is used to interact with a Database instance. Each Connection is thread-safe. Multiple
 * connections can connect to the same Database instance in a multi-threaded environment.
 */
public class Connection implements AutoCloseable {

    long conn_ref;
    boolean destroyed = false;

    /**
     * Creates a connection to the database.
     *
     * @param db: Database instance.
     */
    public Connection(Database db) {
        if (db == null)
            throw new AssertionError("Cannot create connection, database is null.");
        conn_ref = Native.kuzu_connection_init(db);
    }

    /**
     * Check if the connection has been destroyed.
     *
     * @throws ObjectRefDestroyedException If the connection has been destroyed.
     */
    private void checkNotDestroyed() throws ObjectRefDestroyedException {
        if (destroyed)
            throw new ObjectRefDestroyedException("Connection has been destroyed.");
    }

    /**
     * Close the connection and release the underlying resources. This method is invoked automatically on objects managed by the try-with-resources statement.
     *
     * @throws ObjectRefDestroyedException If the connection has been destroyed.
     */
    @Override
    public void close() throws ObjectRefDestroyedException {
        destroy();
    }

    /**
     * Destroy the connection.
     *
     * @throws ObjectRefDestroyedException If the connection has been destroyed.
     */
    private void destroy() throws ObjectRefDestroyedException {
        checkNotDestroyed();
        Native.kuzu_connection_destroy(this);
        destroyed = true;
    }

    /**
     * Return the maximum number of threads used for execution in the current connection.
     *
     * @return The maximum number of threads used for execution in the current connection.
     * @throws ObjectRefDestroyedException If the connection has been destroyed.
     */
    public long getMaxNumThreadForExec() throws ObjectRefDestroyedException {
        checkNotDestroyed();
        return Native.kuzu_connection_get_max_num_thread_for_exec(this);
    }

    /**
     * Sets the maximum number of threads to use for execution in the current connection.
     *
     * @param numThreads: The maximum number of threads to use for execution in the current connection
     * @throws ObjectRefDestroyedException If the connection has been destroyed.
     */
    public void setMaxNumThreadForExec(long numThreads) throws ObjectRefDestroyedException {
        checkNotDestroyed();
        Native.kuzu_connection_set_max_num_thread_for_exec(this, numThreads);
    }

    /**
     * Executes the given query and returns the result.
     *
     * @param queryStr: The query to execute.
     * @return The result of the query.
     * @throws ObjectRefDestroyedException If the connection has been destroyed.
     */
    public QueryResult query(String queryStr) throws ObjectRefDestroyedException {
        checkNotDestroyed();
        return Native.kuzu_connection_query(this, queryStr);
    }

    /**
     * Prepares the given query and returns the prepared statement.
     *
     * @param queryStr: The query to prepare.
     * @return The prepared statement.
     * @throws ObjectRefDestroyedException If the connection has been destroyed.
     */
    public PreparedStatement prepare(String queryStr) throws ObjectRefDestroyedException {
        checkNotDestroyed();
        return Native.kuzu_connection_prepare(this, queryStr);
    }

    /**
     * Executes the given prepared statement with args and returns the result.
     *
     * @param ps: The prepared statement to execute.
     * @param m:  The parameter pack where each arg is a std::pair with the first element being parameter name and second
     *            element being parameter value
     * @return The result of the query.
     * @throws ObjectRefDestroyedException If the connection has been destroyed.
     */
    public QueryResult execute(PreparedStatement ps, Map<String, Value> m) throws ObjectRefDestroyedException {
        checkNotDestroyed();
        return Native.kuzu_connection_execute(this, ps, m);
    }

    /**
     * Interrupts all queries currently executed within this connection.
     *
     * @throws ObjectRefDestroyedException If the connection has been destroyed.
     */
    public void interrupt() throws ObjectRefDestroyedException {
        checkNotDestroyed();
        Native.kuzu_connection_interrupt(this);
    }

    /**
     * Sets the query timeout value of the current connection. A value of zero (the default) disables the timeout.
     *
     * @param timeoutInMs: The query timeout value in milliseconds.
     * @throws ObjectRefDestroyedException If the connection has been destroyed.
     */
    public void setQueryTimeout(long timeoutInMs) throws ObjectRefDestroyedException {
        checkNotDestroyed();
        Native.kuzu_connection_set_query_timeout(this, timeoutInMs);
    }
}

File: tools/wasm/src_js/sync/connection.js
/**
 * @file connection.js is the file for the Connection class. Connection 
 * is used to interact with a Database instance. 
 */
"use strict";

const KuzuWasm = require("./kuzu.js");
const QueryResult = require("./query_result.js");
const PreparedStatement = require("./prepared_statement.js");

class Connection {
  /**
   * Initialize a new Connection object.
   *
   * @param {kuzu.sync.Database} database the database object to connect to.
   * @param {Number} numThreads the maximum number of threads to use for query 
   * execution.
   */
  constructor(database, numThreads = null) {
    KuzuWasm.checkInit();
    const kuzu = KuzuWasm._kuzu;
    if (!database || typeof database !== "object") {
      throw new Error("Database must be an object.");
    }
    if (database._isClosed) {
      throw new Error("Database is already closed.");
    }
    numThreads = parseInt(numThreads);

    this._connection = new kuzu.Connection(database._database);
    if (numThreads && numThreads > 0) {
      this._connection.setMaxNumThreadForExec(numThreads);
    }
    this._isClosed = false;
  }

  /**
   * Internal function to check if the connection is closed.
   * @throws {Error} if the connection is closed.
   * @private
   */
  _checkConnection() {
    KuzuWasm.checkInit();
    if (this._isClosed) {
      throw new Error("Connection is already closed.");
    }
  }

  /**
   * Set the maximum number of threads to use for query execution.
   * @param {Number} numThreads the maximum number of threads to use for query 
   * execution.
   * @throws {Error} if the connection is closed.
   * @throws {Error} if numThreads is not a positive integer.
   */
  setMaxNumThreadForExec(numThreads) {
    this._checkConnection();
    this._connection.setMaxNumThreadForExec(numThreads);
    this._numThreads = numThreads;
  }

  /**
   * Set the query timeout in milliseconds.
   * @param {Number} timeout the query timeout in milliseconds.
   * @throws {Error} if the connection is closed.
   * @throws {Error} if timeout is not a positive integer.
   */
  setQueryTimeout(timeout) {
    this._checkConnection();
    this._connection.setQueryTimeout(timeout);
  }

  /**
   * Get the maximum number of threads to use for query execution.
   * @returns {Number} the maximum number of threads to use for query execution.
   * @throws {Error} if the connection is closed.
   */
  getMaxNumThreadForExec() {
    this._checkConnection();
    return this._connection.getMaxNumThreadForExec();
  }

  /**
   * Execute a query.
   * @param {String} statement the statement to execute.
   * @returns {kuzu.sync.QueryResult} the query result.
   * @throws {Error} if the connection is closed.
   * @throws {Error} if statement is not a string.
   */
  query(statement) {
    this._checkConnection();
    if (typeof statement !== "string") {
      throw new Error("Statement must be a string.");
    }
    const _queryResult = this._connection.query(statement);
    return new QueryResult(_queryResult);
  }

  /**
   * Prepare a statement for execution.
   * @param {String} statement the statement to prepare.
   * @returns {Promise<kuzu.sync.PreparedStatement>} the prepared statement.
   */
  prepare(statement) {
    this._checkConnection();
    if (typeof statement !== "string") {
      throw new Error("Statement must be a string.");
    }
    const _preparedStatement = this._connection.prepare(statement);
    return new PreparedStatement(_preparedStatement);
  }

  /**
   * Execute a prepared statement.
   * @param {kuzu.sync.PreparedStatement} preparedStatement the prepared 
   * statement to execute.
   * @param {Object} params a plain object mapping parameter names to values.
   * @returns {kuzu.sync.QueryResult} the query result.
   * @throws {Error} if the connection is closed.
   * @throws {Error} if preparedStatement is not a valid PreparedStatement 
   * object.
   * @throws {Error} if preparedStatement is not successful.
   * @throws {Error} if params is not a plain object.
   */
  execute(preparedStatement, params = {}) {
    this._checkConnection();
    if (!preparedStatement ||
      preparedStatement._isClosed) {
      throw new Error("preparedStatement is not valid or is closed.");
    }
    if (!preparedStatement.isSuccess()) {
      throw new Error(preparedStatement.getErrorMessage());
    }
    if (!params) {
      throw new Error("params is not valid.");
    }
    const paramsArray = [];
    for (const key in params) {
      paramsArray.push({
        name: key,
        value: params[key]
      });
    }
    const _queryResult = this._connection.execute(preparedStatement._statement, paramsArray);
    return new QueryResult(_queryResult);
  }

  /**
   * Close the connection.
   */
  close() {
    if (!this._isClosed) {
      this._connection.delete();
      this._isClosed = true;
    }
  }
}

module.exports = Connection;
